:art: Raise exception instead of hiding it in finally

gortibaldik opened the pull request:
We use the dependency-injector a lot in our private project. Currently, when the configuration of the ThreadLocalSingleton is invalid (i.e. a constructor argument is forgotten), the library prints only this exception:

```
UnboundLocalError: local variable 'instance' referenced before assignment
```

This is really frustrating, because it provides no leads to the actual error. At the company, for last 7 months, we use a fork of this library, https://github.com/gortibaldik/dependency-injector-fg with a separate PyPI repo. However as we see that this project is again maintained, we would like to merge this small PR, which solves approx 50% of all of our problems with the library (the other 50% are issues with wiring which stem from our misunderstandings... :grinning: ).

gortibaldik commented:
@rmk135, do you have time to look at this ?

On the other hand, we have another small issue. In your pipelines you do not upload wheels to the PyPI, which makes all the builds of our private library too long. In the repo discussed above (https://github.com/gortibaldik/dependency-injector-fg) we push wheels to the PyPI. What do you think about that ?

ZipFile commented on tests/unit/providers/singleton/test_thread_local_singleton_py2_py3.py:
`py2` is extra here, we no longer support Python 2

ZipFile replied:
Also run `isort` and `black` against this file. Code style.

gortibaldik replied:
Hi, fixed both comments.

ZipFile reviewed (commented):
Hey! Looks interesting, I'll check this more in depth later this week.

In the mean time, could you provide a real-world (or something that looks close to it) example of those 50% cases you have issues with? For documentation purposes. Since introduction of contextvars, thread-local data is not something you use often nowadays, so I would like to know your particular use case.

gortibaldik commented:
Are you referring to the 50% use cases where we use ThreadLocalSingleton ? We used the ThreadLocalSingleton for the database (SQLAlchemy), before we switched to the Singleton. Also for some per-thread variables in our FastAPI app.

Due to the current state of ThreadLocalSingleton (with the finally block) whenever the configuration was wrong the "unbound instance" error was raised. Basically anytime we changed per-customer configuration files the error could have risen up.

ZipFile commented:
I see. Try experimenting with `ContextLocalSingleton`, this should work better in combination with FastAPI.

Regarding long build times, I do not think there will be any changes soon. It's quite delicate topic (see #487), so we should be careful here.

I'm personally waiting for 2 things:
* [ARM GHA runenrs](https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/) to be available for opensource repos. ARM builds take forever to build right now due to emulation.
* [Cython 3.1](https://github.com/cython/cython/milestone/65). They promised better [Limited API](https://docs.python.org/3/c-api/stable.html#limited-c-api) support, which may significantly reduce amount of packages we build (hence reduce build time).

gortibaldik commented:
Unfortunately, the project is already in production, and the experimentation is very limited. We likely won't replace ThreadLocalSingleton with ContextLocalSingleton in the forseeable future.

gortibaldik commented:
Hi, what is the state of this MR ? In my opinion it is really small and improves the functionality of the lib, I don't see any reasons to reject or stall this MR. Can you elaborate on this, please ?

rmk135 commented:
@gortibaldik thank you for the PR and thanks everyone here for the tons of useful feedback.

rmk135 commented:
All, @ZipFile and I will push a new release shortly.
